From c54c9563ab72eef6a12a19473a32fd40390eed92 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:46 -0400
Subject: [fs] nfs: fix error handling in lock_and_join_requests

Message-id: <1411050057-12417-84-git-send-email-steved@redhat.com>
Patchwork-id: 96236
O-Subject: [RHEL7.1 PATCH 83/94] nfs: fix error handling in lock_and_join_requests
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Weston Andros Adamson <dros@primarydata.com>

This fixes handling of errors from nfs_page_group_lock in
nfs_lock_and_join_requests.  It now releases the inode lock and the
reference to the head request.

Reported-by: Peng Tao <tao.peng@primarydata.com>
Signed-off-by: Weston Andros Adamson <dros@primarydata.com>
Reviewed-by: Peng Tao <tao.peng@primarydata.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 94970014c46223cbcdfbfc67b89596a412f9e3dd
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/write.c b/fs/nfs/write.c
index 0112e66..eadad96 100644
--- a/fs/nfs/write.c
+++ b/fs/nfs/write.c
@@ -480,8 +480,11 @@ try_again:
 
  /* lock each request in the page group */
  ret = nfs_page_group_lock(head, true);
- if (ret < 0)
+ if (ret < 0) {
+  spin_unlock(&inode->i_lock);
+  nfs_release_request(head);
   return ERR_PTR(ret);
+ }
  subreq = head;
  do {
   /*
-- 
1.7.1